src/Rings/mpoly-graded.jl

```diff
@@ -1562,6 +1562,397 @@ end
+  ##    @req  !is_zero(I)  "Ideal must be non-zero"
+  ##    @req  !is_one(I)   "Ideal must not be whole ring"
+  G = gens(I)
+  if isempty(G)  throw("Ideal must have at least 1 generator"); end;
+  OrigR = parent(G[1])      # Since I is not zero, it has at least 1 gen.
+  # Next few lines: we adjoin some more small, redundant gens.
+  # !!HEURISTIC!!  We use 2*AveNumTerms as size limit for redundant gens we shall adjoin.
+  AveNumTerms = sum([length(g)  for g in G])/length(G) ## floating-point!
+  if extra_gens_flag
+    # compute DegRevLex GB (?and maybe DegLex GB?)
+    assume_gb_is_cached = groebner_basis(I; ordering=degrevlex(OrigR))
+    #??Good idea??        assume_gb_is_cached = groebner_basis(I; ordering=deglex(OrigR))
+  end
+  for GB in values(I.gb)
+    extra_gens = filter(f -> (length(f) < 2*AveNumTerms), elements(GB))
+    G = vcat(G,  extra_gens)
+  end
+  return G
+end
+
+# Fully general, but typically much slower than specialized method below (for positive gradings)
+# Optional kwarg:
+#   extra_gens_flag: if false, inhibits computing grobner basis speculatively in _gens_for_homog_via_sat
+function _homogenization_via_saturation(I::MPolyIdeal{T},  H::homogenizer; extra_gens_flag::Bool = false) where {T <: MPolyRingElem}
+  P = base_ring(I)
+  @req P == H.P  "Homogenizer not for this ring"
+  if is_zero(I)  # special handling for ideal(0)
+    return ideal(zero(H.P_homog)) # return zero ideal in !!P_homog!!
+  end
+  Hgens = H(_gens_for_homog_via_sat(I, extra_gens_flag))
+  R = parent(Hgens[1]) # or H.P_homog
+  if length(Hgens) == 1  # short-cut for principal ideal
+    return ideal(R, Hgens)
+  end
+  DoSatByProduct = false  # true means saturate by product of homogenizing vars;
+                          # false means saturate successively by each homogenizing var (probably faster, in most cases?)
+  HVars = homogenizing_variables(H)
+  if DoSatByProduct
+    Ih = saturation(ideal(R, Hgens), ideal(R, prod(HVars)))
+  else # not DoSatByProduct, so do a cascade of saturations
+    Ih = ideal(R, Hgens)
+    for hvar in HVars
+      Ih = saturation(Ih, ideal(R, hvar))
+      # IMPROVE???  if length(gens(Ih)) == 1  return Ih;  end
+    end
+  end
+  return Ih
+end
+
+
+# ============================================
+# 2023-06-30 START: New impl of homogenization
+# By John Abbott, based on K+R Book "Computational Commutative Algebra" (see "justification" below)
+# This code delegates to _homogenization_via_saturation for non-positive gradings
+# ============================================
+
+# _sat_poly_by_var and kronecker_delta are internal auxiliary functions
+# This homogenization impl seems to be usefully faster with positive ZZ^m-grading for m > 1
+
+# Saturate a polynomial by a variable -- equiv to  f/gcd(f,h^deg(f))
+function _sat_poly_by_var(f,h)
+  # ASSUMES h is a variable i.e. monic poly of deg 1 with just 1 term
+  # Is there a better way to implement this?
```

Comment:
There is the function `remove`:
```
  remove(z::MPolyRingElem{T}, p::MPolyRingElem{T}) where {T <: RingElement}

  Compute the valuation of z at p, that is, the largest k such that p^k divides z. Additionally, z/p^k
  is returned as the second return. value.

  See also valuation, which only returns the valuation.
```
I don't know whether it is better though :)

Reply:
Ahh.  After a little testing, neither my implementation nor that of `remove` is well-suited to this use-case (though both are mathematically correct).

Reply:
I have written specific code, which is faster.